### PR TITLE
Adding first version of authentication and authorization

### PR DESCRIPTION
**Pull request by jlarfors**
First draft of authn and authz.

There are some things to consider:
1. This doesn't actually change anything unless you opt in for authentication and multitenancy
2. How do the conditional routes (e.g with multitenancy the base API route is `/api/v1/o/:organization`) affect swagger and the Bubbly CLI?

Open to feedback and suggestions!

And I also accidentally through in some changes to the server that were bothering me... Basically adding `bCtx` to the `server` so that it doesn't need to get passed around everywhere

### CODE REVIEW

File: server/routes.go [hunk 1]
```diff
@@ -6,44 +6,42 @@ import (
...
-
-		api.POST("/schema", func(c echo.Context) error {
-			return s.PostSchema(bCtx, c)
-		})
-
-		api.POST("/upload", func(c echo.Context) error {
-			return s.upload(bCtx, c)
-		})
+
+	// If multitenancy is enabled, add it the URL path
```
Comment (nate-droid) [inline server/routes.go:20]: Really nice and clean! Love it.

File: server/middleware.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+		if resp.StatusCode != http.StatusOK {
+			return echo.NewHTTPError(resp.StatusCode, "authorization request failed")
+		}
+
+		var reqAuth requestAuth
+		err = json.NewDecoder(resp.Body).Decode(&reqAuth)
+		if err != nil {
+			return echo.NewHTTPError(http.StatusInternalServerError, "failed getting request authentication: "+err.Error())
+		}
+		// fmt.Printf("\n\nWHAAAAT! It worked!\n%v\n", reqAuth)
```
Comment (nate-droid) [inline server/middleware.go:64]: I love the surprise 😆

File: server/auth.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
+package server
+
+import (
+	"io/ioutil"
+	"net/http"
+
+	"github.com/labstack/echo/v4"
+)
+
+func (s *Server) authorize(c echo.Context) error {
```
Comment (nate-droid) [inline server/auth.go:10]: I really like this approach as well. Nice and clean!

**Review by nate-droid — APPROVED** [review]

File: server/middleware.go (hunk 2 above)
Comment (jlarfors) [reply server/middleware.go]: Oops... it's fixed!

**Comment by benmarsden** [comment]
> 2\. How do the conditional routes (e.g with multitenancy the base API route is `/api/v1/o/:organization`) affect swagger and the Bubbly CLI?

I think this may break the CLI: the CLI makes requests through the bubbly HTTP Client, which has static routes defined like so:

```go
// client/query.go
http.Post(fmt.Sprintf("%s/api/v1/graphql", c.HostURL), "application/json", bytes.NewBuffer(jsonReq)),
```

So I think you'll need to update this request logic to accommodate the new base route.

I do not have the knowledge to comment on swagger.

**Review by benmarsden — CHANGES_REQUESTED** [review]
Update Bubbly's HTTP Client to handle differing base routes.

**Review by olliefr — CHANGES_REQUESTED** [review]
Just some Swagger-related stuff... 🙃

File: .github/workflows/integration.yaml [hunk 4]
```diff
@@ -25,7 +25,7 @@ jobs:
             ${{ runner.os }}-go-
       - name: Integration tests
         run: |
-          docker-compose up -d cockroachdb bubbly
+          docker-compose up -d postgres bubbly
```
[the remark this answers is not in the document]
Comment (jlarfors) [reply .github/workflows/integration.yaml:28]: Yup, fair point...

File: server/auth.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return nil, echo.NewHTTPError(resp.StatusCode, "received non-200 HTTP status code from auth service")
+	}
+
+	// Read and forward the bytes
+	respBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, echo.NewHTTPError(http.StatusInternalServerError, "failed to read response from auth service: "+err.Error())
```
[the remark this answers is not in the document]
Comment (jlarfors) [reply server/auth.go:110]: Nope, it shouldn't

**Review by olliefr — APPROVED** [review]

**Review by nate-droid — APPROVED** [review]

**Pull request merged**